File: src/main/java/me/pcasaes/hexoids/core/domain/model/DomainEvent.java
package me.pcasaes.hexoids.core.domain.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import pcasaes.hexoids.proto.Event;

import java.util.UUID;

public class DomainEvent {
    private final UUID key;
    private final Event event;
    private final String topic;

    private DomainEvent(String topic, UUID key, Event event) {
        this.topic = topic;
        this.key = key;
        this.event = event;
    }


    public static DomainEvent of(
            UUID key,
            Event event) {
        return new DomainEvent(null, key, event);
    }

    public static DomainEvent create(String topic, UUID key, Event event) {
        return new DomainEvent(topic, key, event);
    }

    public static DomainEvent delete(String topic, UUID key) {
        return new DomainEvent(topic, key, null);
    }

    public static DomainEvent deleted(UUID key) {
        return new DomainEvent(null, key, null);
    }

    public UUID getKey() {
        return key;
    }

    public Event getEvent() {
        return event;
    }

    @JsonIgnore
    public String getTopic() {
        return topic;
    }
}

File: src/main/kotlin/me/pcasaes/hexoids/core/domain/model/DomainEvent.kt
package me.pcasaes.hexoids.core.domain.model

import com.fasterxml.jackson.annotation.JsonIgnore
import pcasaes.hexoids.proto.Event
import java.util.UUID

class DomainEvent private constructor(
    @get:JsonIgnore val topic: String?,
    val key: UUID,
    val event: Event?
) {

    companion object {

        fun of(
            key: UUID,
            event: Event?
        ): DomainEvent {
            return DomainEvent(null, key, event)
        }

        fun create(topic: String, key: UUID, event: Event): DomainEvent {
            return DomainEvent(topic, key, event)
        }

        fun delete(topic: String, key: UUID): DomainEvent {
            return DomainEvent(topic, key, null)
        }

        fun deleted(key: UUID): DomainEvent {
            return DomainEvent(null, key, null)
        }
    }
}
